New OCD export

dg0yt opened the pull request:
Will eventually resolve #868, #596.

Covered versions: 8, 9, 10, 11, 12
- [ ] version 12 testing

Covered features:
- [x] georeferencing - needs #1034
- [x] regular colors
- [x] spot colors
- [x] V6 symbol icons (16 colors)
- [x] V9 symbol icons (125 colors)
- [x] point symbols
- [x] line symbols
- [x] area symbols
- [x] text symbols
- [x] combined symbols
  - [x] combined symbols for special OCD line and area symbols
- [x] point objects
- [x] path objects
- [x] text objects
- [ ] line text symbols/objects - needs new symbol type/object type
- [x] map notes
- [x] map view
- [x] templates
- [ ] other data

lpechacek commented:
Good! I have the georeferencing translation ready. I'll look into plugging it into the new export.

dg0yt commented:
Now rebased on dev to pull in namespace OpenOrienteering.

dg0yt commented:
- Import and export features are fairly balanced in the modern implementations. I wonder if I should either  completely disable the old import/export code (for simplicity). Otherwise I should at least disable mixed use.
- How to select an OCD version? I could use the imported version when saving to the same file, and ask otherwise. However, asking questions might need new code. Another option is to use a default version from the settings.
- I will merge the georeferencing PR #1034 as soon as the last comments are dealt with.

mstahv commented:
Good job Kai and other developers, this will be a game changer in the adoption of Mapper!

lpechacek commented:
Proposed integration between georef and export is at https://github.com/lpechacek/mapper/commit/50dbfee2a7a9dbb287df0ee4d2e5a87d6c99cb84. Will create PR with this commit when #1034 is finished.

lpechacek commented:
> Import and export features are fairly balanced in the modern implementations. I wonder if I should either completely disable the old import/export code (for simplicity). Otherwise I should at least disable mixed use.

Keep the old code aside, guarded by a run time setting, so that users can switch to it if the new implementation exhibits bugs. My impression is that it was done so for the GDAL based importers.

> How to select an OCD version? I could use the imported version when saving to the same file, and ask otherwise. However, asking questions might need new code. Another option is to use a default version from the settings.

Present different versions as different file types during export:
![image](https://user-images.githubusercontent.com/5584406/37700566-52313634-2cec-11e8-89f0-71777abe8468.png)

dg0yt commented:
> Keep the old code aside, guarded by a run time setting, so that users can switch to it if the new implementation exhibits bugs. My impression is that it was done so for the GDAL based importers.

With GDAL, it was a one way road anyway. The new export is balanced with changes in the import. When using the new import with the old export, the result will be worse than saving to Mapper format first, and then loading and saving to OCD with an older release of Mapper.

> Present different versions as different file types during export:

Possible indeed. The [selected filter is exposed](https://doc.qt.io/qt-5/qfiledialog.html#selectedNameFilter) but I need to verify that this works cross platform with the native dialogs and with translations. So far we only on the extension.

dg0yt commented:
Fixed the FileFormatTest, and updated version selection on saving.

What I would like to know from review at the moment is whether this together with #1081 is mature enough for merging to `dev`, making it available for wider testing via the unstable builds. (I would announce that on openorienteering.org.)

The georeferencing PR #1034 still needs changes ATM.

lpechacek commented:
> What I would like to know from review at the moment is...

I'll do my best to look into it this week.

lpechacek commented:
First feedback from trying out the new export. First and foremost, good job done! :)

I tried exporting a map I had at hand, opened it in OCAD 12 Viewer running under Wine and spotted small differences in symbol rendering. There are variations in pattern clipping and rendering and dash placement. I created a sample file illustrating differences I see. No objects were lost during export.

Regarding the export process itself, I considered _Save as..._ file format descriptions "OCAD" and "OCAD legacy export" insufficiently descriptive. How about adding version number to these options and extending the plain "OCAD" with some adjective?

[different-rendering.zip](https://github.com/OpenOrienteering/mapper/files/1857411/different-rendering.zip)

lpechacek commented:
Regarding code maturity, I propose [fuzzing](https://llvm.org/docs/LibFuzzer.html) the import/export code. That will surely shake out quite some bugs.

dg0yt commented:
> Regarding the export process itself, I considered Save as... file format descriptions "OCAD" and "OCAD legacy export" insufficiently descriptive. How about adding version number to these options and extending the plain "OCAD" with some adjective?

That's #1081, isn't it?

dg0yt commented:
> I tried exporting a map I had at hand, opened it in OCAD 12 Viewer running under Wine and spotted small differences in symbol rendering. There are variations in pattern clipping and rendering and dash placement. I created a sample file illustrating differences I see. No objects were lost during export.

OCAD 12 Viewer doesn't run with my Wine installation, so I couldn't test the latest features of the format. Minor difference are expected anyway, as the rendering is performed by different implementations. That's what needs to be polished slowly. What I can do now is take the sample map from OCAD 12 View, save as OCAD 11, and open in OCAD 11 Viewer :-)

The core of the PR is the ability to handle all the different version from a small set of template functions and C++ style API, very different from the previous libocad C library for version 8 only. It is just the base for future work on quality.

lpechacek commented:
> That's #1081, isn't it?

I merged both PR's into one branch for my testing to get the feeling how Mapper will look like after these two patch sets hit origin/dev. Anyway, I'll have a closer look at PR#1081 to see what role it plays.

lpechacek commented on src/fileformats/ocd_file_export.cpp:
```diff
@@ -48,4 +884,2019 @@ void OcdFileExport::doExport()
+
+
+template< class OcdAreaSymbolCommon >
+quint8 OcdFileExport::exportAreaSymbolCommon(const AreaSymbol* area_symbol, OcdAreaSymbolCommon& ocd_area_common, const PointSymbol*& pattern_symbol)
```
Long parameter list does not fit onto screen. How about breaking the line?

lpechacek commented on src/fileformats/ocd_file_import.cpp:
```diff
@@ -439,21 +565,146 @@ void OcdFileImport::importColors(const OcdFile<Ocd::FormatV8>& file)
+	}
+}
+
+void OcdFileImport::importSpotColor(const QString& param_string)
```
Glad to see spot colors being handled now!

lpechacek commented on src/fileformats/ocd_file_import.h:
```diff
@@ -350,114 +352,11 @@ class OcdFileImport : public Importer
-
-
-
-// ### OcdFileImport inline code ###
```
A welcome header file cleanup. Thanks!

dg0yt commented on src/fileformats/ocd_file_export.cpp:
```diff
@@ -48,4 +884,2019 @@ void OcdFileExport::doExport()
+
+template<class Format>
+void OcdFileExport::exportSetup(OcdFile<Format>& /*file*/)
+{
```
Because `exportSetup()` can do it, and `exportSymbols()`can't.

The fairly consistent pattern in the implementation of `OcdFileExport` (and import): 
From `template<class Format> void OcdFileExport::exportImplementation()` we call into template functions so that we get a template function instantiation for the actual types of the format. This is what we can do:
- If a format is significantly different, so that it needs a different algorithm:
  - We specialize the template function (as done in `template<> void OcdFileExport::exportSetup(OcdFile<Ocd::FormatV8>& file)`).
- If the same algorithm can be used for multiple formats:
  - If the differences are completely represented by the the exporter's state (by capturing version and parameter string handler), then we may just call into a non-template function (as done `template<class Format> void OcdFileExport::exportSetup(OcdFile<Format>& /*file*/)`. This template function is likely to get inlined, thus reducing/avoiding template bloat.
  - Otherwise we still can use the same algorithm but need to specialize it for the actual types of the format. This specialization is solved by function template instantiation, i.e. the generic algorithm goes directly into the template function. This is the case for `exportSymbol` where types and even the encoding of the symbol number vary over different versions.
  This doesn't mean that there couldn't be room for reducing template bloat by factoring out universal parts.

The basic course of development is to start with generic functions (to study feasability), to add test code (to maintain correctness), and to refactor from template functions into regular functions (to reduce template bloat).

lpechacek replied:
Got it. Information about the implementation plan was missing for full understanding the structure.

dg0yt commented on src/fileformats/ocd_file_export.cpp:
```diff
@@ -48,4 +884,2019 @@ void OcdFileExport::doExport()
+	
+	switch (std::extent<typename std::remove_pointer<decltype(ocd_base_symbol.icon_bits)>::type>::value)
+	{
+	case 264:
```
We need to create explicit types representing the icon data. Then we can just call a polymorphic `exportSymbolIcon()`, without `switch` statement.

lpechacek replied:
Agreed. A `\todo` should be added until type-based handling is implemented.

dg0yt commented on src/fileformats/ocd_file_export.cpp:
```diff
@@ -22,22 +22,858 @@
+
+constexpr qint32 convertPointMember(qint32 value)
+{
+	return (value < -5) ? qint32(0x80000000u | ((0x7fffffu & quint32((value-4)/10)) << 8)) : qint32((0x7fffffu & quint32((value+5)/10)) << 8);
```
This function handles two responsibilites at the same time, in a `constexpr` implementation:
1. shift by 8 bits (which are reserved for flags in OCD format)
2. convert from 1/100 mm to to 1/10 mm, rounding half up (for intervalls of equal size).

Neither shifting of signed integers ("implementation-defined") nor rounding (the result of an integer division) half up come out of the box in C++.

I will see how I can put this into a source code comment.

dg0yt commented on src/fileformats/ocd_file_export.cpp:
```diff
@@ -22,22 +22,858 @@
+	    { 255, 255, 255 },
+	};
+	
+#if 0
```
It tells you how the magical table above was created (how it can be changed).

I'm fine with moving it out once the table is accepted. However, I will have to invent a way to organize are generators and tools (`SymbolSetTool`, which is a useful test and a generator at the same time.)

lpechacek replied:
Understood.

lpechacek commented:
As a side note, I've just exported a real map into v12 format and opened it with Viewer 12. While playing with the various menu items I found out:
* Spot colors have all the same angle and frequency. (IIRC a known issue)
* Running Optimize/Repair reported about a fifth of the objects as "damaged" and removed them form the map, substantially changing the map work appearance. It is unclear at the moment why Viewer considered the objects damaged.

The two items should either be dealt with within this PR or moved over to Issues.

dg0yt commented on src/fileformats/ocd_file_export.cpp:
```diff
@@ -48,4 +945,2020 @@ void OcdFileExport::doExport()
+			break;
+			
+		case AreaSymbol::FillPattern::PointPattern:
+		    switch (ocd_area_common.hatch_mode)
```
As you noticed, the wrong variable is used here. I wonder if this could be turned in a compile-time error.

However, `structure_mode` is not really used as counter, i.e. it is not incremented, but explicitly switched from one state to another as needed to construct a representation of the pattern.

lpechacek replied:
> As you noticed, the wrong variable is used here. I wonder if this could be turned in a 
> compile-time error.

If nothing else, the following patch makes Clang 6 emit a warning.

[typed-symbol-area-modes.txt](https://github.com/OpenOrienteering/mapper/files/1947422/typed-symbol-area-modes.txt)

dg0yt commented:
We need more changes related to file formats, importers and exporters to handle OCD versions properly fir opening, saving and save-as. I will try to do this separately before coming back to the last WIP commit.

lpechacek commented:
FWIW, I've successfully used this work for exporting a map with 2700 objects into OCAD 11 format. There are surely nits here and there but in general the new export works. I'm for publishing it as an Alpha in the hope that other users will happily join testing.

As a side note, there is a simple procedure involving Gimp for checking that original map and exported map look the same. Good night.
